### PR TITLE
mgr/cephadm: nvme-of follow up work

### PR DESCRIPTION
**Pull request by adk3798**
The first piece  of the nvme-of deployment work in cephadm was already merged in main. This PR is a follow up that fixes a handful of issues found in the original implementation.

This depends on having an nvme-of container with changes being proposed in https://github.com/ceph/ceph-nvmeof/pull/166. For now, I have a temporary commit that points the default nvme-of image to an image on my personal quay with the changes present. That should be removed before merging (signed-off-by will fail as long as that commit is present.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/services/nvmeof.py
Comment (phlogistonjohn) [inline src/pybind/mgr/cephadm/services/nvmeof.py:37]: Probably unrelated to this change, but it might be good to fix this up sooner rather than later, but the var name `igw_id` makes no sense to me.

File: src/pybind/mgr/cephadm/serve.py
Comment (phlogistonjohn) [inline src/pybind/mgr/cephadm/serve.py:923]: Commit message typo: `logicas` should be `logic as`

**Review by phlogistonjohn — COMMENTED** [review]
Generally LGTM, happy to approve once PR is out of WIP state. A few minor feedback items follow.

File: qa/suites/orch/cephadm/smoke-roleless/2-services/nvmeof.yaml
Comment (adk3798) [reply qa/suites/orch/cephadm/smoke-roleless/2-services/nvmeof.yaml:5]: Very little, if anything at all. One "initializes" a pool for rbd use and the other lets you "enable" the use of rbd on the pool (from their descriptions). I think we can use them interchangeably in this context.

File: src/pybind/mgr/cephadm/services/nvmeof.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/services/nvmeof.py:37]: yeah, I think that's leftover from the nvmeof implementation being mostly copied from the iscsi one. I can add a commit to rename the variable as well.

**Comment by phlogistonjohn** [comment]
The current changes LGTM, modulo the temporary container image hack. Once there's a usable upstream build with the change and you drop that temporary commit please ping me for final approval!

File: qa/suites/orch/cephadm/upgrade/3-upgrade/simple.yaml
Comment (adk3798) [inline qa/suites/orch/cephadm/upgrade/3-upgrade/simple.yaml:17]: the normal nvmeof deployment test worked, but this upgrade one won't because the upgrade starts from pacific (need to update this to start from quincy actually since we're now in the dev cycle for squid, but that won't help this) and we can't deploy nvmeof on older versions.

Comment (phlogistonjohn) [reply qa/suites/orch/cephadm/upgrade/3-upgrade/simple.yaml:17]: That sounds reasonable to me.

**Comment by adk3798** [comment]
> The current changes LGTM, modulo the temporary container image hack. Once there's a usable upstream build with the change and you drop that temporary commit please ping me for final approval!

this is no longer pointing towards a temp image. An official nvmeof image 0.0.2 was made with the necessary changes (https://github.com/ceph/ceph-nvmeof/issues/172) and we now point to that.

**Review by phlogistonjohn — APPROVED** [review]
LGTM

**Comment by adk3798** [comment]
jenkins retest this please

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-08-15_13:44:09-orch:cephadm-wip-adk-testing-2023-08-14-1902-distro-default-smithi/

7 failures, 3 dead (as of writing this 1 dead and 2 running but the 2 running will time out) jobs

- 4 failures were mds_upgrade_sequence upgrading from reef -> main. The upgrade from reef doesn't work yet as we need to implement something that can pull the latest reef build of the cephadm binary
- 2 failures were https://tracker.ceph.com/issues/62084, known issue
- 1 failure was https://tracker.ceph.com/issues/59704, known issue
- 2 dead jobs were mds upgrade sequence starting from quincy. They time out doing a `umount` command. These were broken the same way before when upgrading from pacific, and we wanted to see if upgrading from quincy would fix it somehow but it hasn't.
- 1 dead job was infra/teuthology issue: `Error reimaging machines: Expected smithi172's OS to be centos 8 but found rhel 8.6`

Overall, nothing related to the mds upgrade sequence or  jaeger-tracing can be merged, but nothing here to block merging most other things in the run.

**Pull request merged**
